esmf_regrid/experimental/unstructured_scheme.py

```diff
@@ -0,0 +1,152 @@
+"""Provides an iris interface for unstructured regridding."""
+
+import iris
+from iris.analysis._interpolation import get_xy_coords
+
+# from numpy import ma
+
+from esmf_regrid.esmf_regridder import Regridder
+
+# from esmf_regrid.esmf_regridder import GridInfo
+# from esmf_regrid.experimental.unstructured_regrid import MeshInfo
+
+
+def _get_mesh_and_dim(cube):
+    # Returns the cube's mesh and the dimension that mesh belongs to.
+    # Likely to be of the form:
+    # mesh = cube.mesh
+    # mesh_dim = cube.mesh_dim(mesh)
+    # return mesh, mesh_dim
+    pass
+
+
+def _cube_to_MeshInfo(cube):
+    # Returns a MeshInfo object describing the mesh of the cube.
+    pass
+
+
+def _cube_to_GridInfo(cube):
+    # Returns a GridInfo object describing the horizontal grid of the cube.
+    # This may be inherited from code written for the rectilinear regridding scheme.
+    pass
+
+
+# def _regrid_along_dims(regridder, data, src_dim, mdtol):
+#     # Before regridding, data is transposed to a standard form.
+#     # This will be done either with something like the following code
+#     # or else done within the regridder by specifying args.
+#     # new_axes = list(range(len(data.shape)))
+#     # new_axes.pop(src_dim)
+#     # new_axes.append(src_dim)
+#     # data = ma.transpose(data, axes=new_axes)
+#
+#     result = regridder.regrid(data, mdtol=mdtol)
+#     return result
+
+
+def _create_cube(data, src_cube, mesh_dim, mesh, grid_x, grid_y):
+    # Here we expect the args to be as follows:
+    # data: a masked array containing the redult of the regridding operation
+    # src_cube: the source cube which data is regrid from
+    # mesh_dim: the dimension on src_cube which the mesh belongs to
+    # mesh: the Mesh (or MeshCoord) object belonging to src_cube
+    # grid_x: the coordinateon the target cube representing the x axis
+    # grid_y: the coordinateon the target cube representing the y axis
+
+    new_cube = iris.cube.Cube(data)
+
+    # TODO: add coords and metadata.
+
+    return new_cube
+
+
+def _regrid_unstructured_to_rectilinear__prepare(src_mesh_cube, target_grid_cube):
+    # TODO: Perform checks on the arguments. (grid coords are contiguous,
+    #  spherical and monotonic. Mesh is defined on faces)
+
+    # TODO: Account for differences in units.
+
+    # TODO: Account for differences in coord systems.
+
+    # TODO: Record appropriate dimensions (i.e. which dimension the mesh belongs to)
+
+    grid_x, grid_y = get_xy_coords(target_grid_cube)
+
+    meshinfo = _cube_to_MeshInfo(src_mesh_cube)
+    gridinfo = _cube_to_GridInfo(target_grid_cube)
+
+    regridder = Regridder(meshinfo, gridinfo)
+
+    mesh, mesh_dim = _get_mesh_and_dim(src_mesh_cube)
+
+    regrid_info = (mesh, mesh_dim, grid_x, grid_y, regridder)
```

Comment:
For my own understanding, there is a possibility that the exact variables in this list are altered once we continue the implementation.  However, they are good for now.